vm/actor/src/builtin/multisig/mod.rs

```diff
@@ -292,105 +362,208 @@ impl Actor {
+                return Err(actor_error!(ErrForbidden;
+                        "{} already approved this message", previous_approver));
+            }
+        }
 
-                // update approved on the transaction
-                txn.approved.push(from);
+        let st = rt.transaction::<State, Result<_, ActorError>, _>(|st, rt| {
+            let mut ptx = make_map_with_root(&st.pending_txs, rt.store()).map_err(
+                |e| actor_error!(ErrIllegalState; "failed to load pending transactions: {}", e),
+            )?;
 
-                if let Err(e) = st.put_pending_transaction(rt.store(), tx_id, txn.clone()) {
-                    return Err(ActorError::new(
-                        ExitCode::ErrIllegalState,
-                        format!("Failed to put transaction for approval: {}", e),
-                    ));
-                }
+            // update approved on the transaction
+            txn.approved.push(*rt.message().caller());
 
-                // Check if number approvals is met
-                if txn.approved.len() >= st.num_approvals_threshold as usize {
-                    // Ensure sufficient funds
-                    if let Err(e) = st.check_available(curr_bal, txn.value.clone(), curr_epoch) {
-                        return Err(ActorError::new(
-                            ExitCode::ErrInsufficientFunds,
-                            format!("Insufficient funds unlocked: {}", e),
-                        ));
-                    }
-
-                    // Delete pending transaction
-                    if let Err(e) = st.delete_pending_transaction(rt.store(), tx_id) {
-                        return Err(ActorError::new(
-                            ExitCode::ErrIllegalState,
-                            format!("failed to delete transaction for cleanup: {}", e),
-                        ));
-                    }
-
-                    Ok((txn, true))
-                } else {
-                    // Number of approvals required not met, do not relay message
-                    Ok((txn, false))
-                }
-            })??;
+            ptx.set(tx_id.key(), txn.clone()).map_err(|e| {
+                actor_error!(ErrIllegalState;
+                    "failed to put transaction {} for approval: {}", tx_id.0, e)
+            })?;
+
+            st.pending_txs = ptx.flush().map_err(
+                |e| actor_error!(ErrIllegalState; "failed to flush pending transactions: {}", e),
+            )?;
 
-        // Sufficient number of approvals have arrived, relay message
-        if threshold_met {
-            rt.send(tx.to, tx.method, tx.params, tx.value)?;
+            // Go implementation holds reference to state after transaction so this must be cloned
+            // to match to handle possible exit code inconsistency
+            Ok(st.clone())
+        })??;
+
+        execute_transaction_if_approved(rt, &st, tx_id, txn)
+    }
+}
+
+fn execute_transaction_if_approved<BS, RT>(
```

Comment:
This fn seems really weird to me, I know its 1:1 matching. It just seems weird that we would be returning an Ok exit code most of the time and wouldn't `applied` always be `true`?

Reply:
well the exit code is directly used in a serialized data structure, so I don't see the benefit in not returning the exit code (or how else would you pass that info back, because a result within a result doesn't seem like a good option.

And no applied is only true if the multisig threshold is met and that block of code gets executed

Reply:
I see, so if the multisig threshold is NOT met we still return an Ok exit code even though applied would be false?

Reply:
yeah, they leave as uninitialized, which is the 0 case (ExitCode::Ok) and return that (I think it's a bit weird that they don't make the exit code and return nullable, but I assume this may be more consistent and less likely to panic if things are changed)